capsules/src/net/icmpv6/icmpv6.rs

```diff
@@ -0,0 +1,182 @@
+            ICMP6Type::Type129 => ICMP6HeaderOptions::Type129 { id: 0, seqno: 0 },
+        };
+
+        ICMP6Header {
+            code: 0,
+            cksum: 0,
+            options: options,
+            len: 0,
+        }
+    }
+
+    pub fn set_type(&mut self, icmp_type: ICMP6Type) {
+        match icmp_type {
+            ICMP6Type::Type1 => self.set_options(ICMP6HeaderOptions::Type1 { unused: 0 }),
+            ICMP6Type::Type3 => self.set_options(ICMP6HeaderOptions::Type3 { unused: 0 }),
+            ICMP6Type::Type128 => self.set_options(ICMP6HeaderOptions::Type128 { id: 0, seqno: 0 }),
+            ICMP6Type::Type129 => self.set_options(ICMP6HeaderOptions::Type129 { id: 0, seqno: 0 }),
+        }
+    }
+
+    pub fn set_code(&mut self, code: u8) {
+        self.code = code;
+    }
+
+    pub fn set_cksum(&mut self, cksum: u16) {
+        self.cksum = cksum;
+    }
+
+    pub fn set_options(&mut self, options: ICMP6HeaderOptions) {
+        self.options = options;
+    }
+
+    pub fn set_len(&mut self, len: u16) {
+        self.len = len;
+    }
+
+    pub fn get_type(&self) -> ICMP6Type {
+        match self.options {
+            ICMP6HeaderOptions::Type1 { .. } => ICMP6Type::Type1,
+            ICMP6HeaderOptions::Type3 { .. } => ICMP6Type::Type3,
+            ICMP6HeaderOptions::Type128 { .. } => ICMP6Type::Type128,
+            ICMP6HeaderOptions::Type129 { .. } => ICMP6Type::Type129,
+        }
+    }
+
+    pub fn get_type_as_int(&self) -> u8 {
+        match self.get_type() {
+            ICMP6Type::Type1 => 1,
+            ICMP6Type::Type3 => 3,
+            ICMP6Type::Type128 => 128,
+            ICMP6Type::Type129 => 129,
+        }
+    }
+
+    pub fn get_code(&self) -> u8 {
+        self.code
+    }
+
+    pub fn get_cksum(&self) -> u16 {
+        self.cksum
+    }
+
+    pub fn get_options(&self) -> ICMP6HeaderOptions {
+        self.options
+    }
+
+    pub fn get_len(&self) -> u16 {
+        return self.len;
+    }
+
+    pub fn get_hdr_size(&self) -> usize {
+        return 8;
+    }
+
+    pub fn encode(&self, buf: &mut [u8], offset: usize) -> SResult<usize> {
+        let mut off = offset;
+
+        off = enc_consume!(buf, off; encode_u8, self.get_type_as_int());
+        off = enc_consume!(buf, off; encode_u8, self.code);
+        off = enc_consume!(buf, off; encode_u16, self.cksum);
+
+        match self.options {
+            ICMP6HeaderOptions::Type1 { unused } | ICMP6HeaderOptions::Type3 { unused } => {
+                off = enc_consume!(buf, off; encode_u32, unused);
+            }
+            ICMP6HeaderOptions::Type128 { id, seqno }
+            | ICMP6HeaderOptions::Type129 { id, seqno } => {
+                off = enc_consume!(buf, off; encode_u16, id);
+                off = enc_consume!(buf, off; encode_u16, seqno);
+            }
+        }
+
+        stream_done!(off, off);
+    }
+
+    pub fn decode(buf: &[u8]) -> SResult<ICMP6Header> {
```

Comment:
It's pretty critical to document that `encode/decode` actually handle _part_ of of the payload (namely the header payload).